File: ncarg2d/src/libncarg/ngmisc/ngsrat.f
C
C       $Id: ngsrat.f,v 1.4 2008-07-27 00:17:18 haley Exp $
C                                                                      
C                Copyright (C)  2000
C        University Corporation for Atmospheric Research
C                All Rights Reserved
C
C The use of this Software is governed by a License Agreement.
C
      SUBROUTINE NGSRAT(IOPT, IAT, RAT)
C
C Save/restore/set NCAR GKS attributes.
C
C ARGUMENTS
C
C ON INPUT  
C    
C   IOPT  An integer argument indicating the desired action:
C
C           = 0, then save the current settings of all attributes 
C                internal to the subroutine, i.e. not in the IAT and
C                RAT arrays.
C           = 1, then restore the context from the most recent
C                saved state.
C           = 2, then return the current context in the output arrays
C                IAT (integer attributes) and RAT (floating-point
C                attributes) as follows:
C
C                IAT - An array containing GKS integer state variables 
C                      as follows:
C
C                        IAT( 1) = Clip indicator
C                        IAT( 2) = Line type
C                        IAT( 3) = Polyline color index
C                        IAT( 4) = Marker type
C                        IAT( 5) = Polymarker color index
C                        IAT( 6) = Text font
C                        IAT( 7) = Text precision
C                        IAT( 8) = Text color index
C                        IAT( 9) = Text path
C                        IAT(10) = Text horizontal alignment
C                        IAT(11) = Text vertical alignment
C                        IAT(12) = Fill area interior style
C                        IAT(13) = Fill are style index
C                        IAT(14) = Fill area color index
C
C                RAT - A REAL array containing GKS floating-point
C                      state variables as follows:
C
C                        RAT( 1) = Linewidth scale factor
C                        RAT( 2) = Marker scale factor
C                        RAT( 3) = Character expansion factor
C                        RAT( 4) = Character spacing
C                        RAT( 5) = Character height in world coordinates
C                        RAT( 6) = Character up vector, X component in
C                                  world coordinates
C                        RAT( 7) = Character up vector, Y component in
C                                  world coordinates
C 
C           = 3, then set the context to the values specified
C                in the IAT and RAT arrays (as described above).
C
C  IAT   An integer array that has meaning only if IOPT equals 2 or 3
C        as indicated above in the description of IOPT.
C
C  RAT   A real array that has meaning only if IOPT equals 2 or 3
C        as indicated above in the description of IOPT.
C
C ---------------------------------------------------------------------------
C  Modified 1/2012 in support of 32bit argb color -- RLB
C
C  Argb colors are generally very large integer values, much bigger
C  than the 5-character field originally allotted for conversion to/from
C  integers.  For simplicity, changed all integer character-fields to
C  10 characters; note that this change necessitated writing across
C  4 80-character records (i.e, IDR), rather than 3.
C ---------------------------------------------------------------------------

      DIMENSION IAT(14),RAT(7)
      CHARACTER*80 IDR(4),ODR(4)
C
      WRITE(IDR(1)(1:5),500) IOPT
  500 FORMAT(I5)
      IF (IOPT.EQ.0 .OR. IOPT.EQ.1) THEN
        CALL GESC(-1388,1,IDR,4,4,ODR)
      ELSE IF (IOPT .EQ. 2) THEN
        CALL GESC(-1388,1,IDR,4,4,ODR)
CRLB:   READ (ODR(1),510) (IAT(LL),LL=1,14)
        READ (ODR(1), 511) (IAT(LL),LL=1,7)
        READ (ODR(2), 511) (IAT(LL),LL=8,14)
  510   FORMAT(14I5)
  511   FORMAT(7I10)
        READ (ODR(3),520) (RAT(LL),LL=1,5)
  520   FORMAT(5E16.7)
        READ (ODR(4),530) (RAT(LL),LL=6,7)
  530   FORMAT(2E16.7)
      ELSE IF (IOPT.EQ.3) THEN
CRLB:   WRITE(IDR(1)(6:75),510) (IAT(LL),LL=1,14)
        WRITE(IDR(1)(6:75),511) (IAT(LL),LL=1,7)
        WRITE(IDR(2),511) (IAT(LL),LL=8,14)
        WRITE(IDR(3),520) (RAT(LL),LL=1,5)
        WRITE(IDR(4),530) (RAT(LL),LL=6,7)
        CALL GESC(-1388,1,IDR,4,4,ODR)
      ENDIF
C
      RETURN
      END

File: ncarg2d/man/ngmisc/ngsrat.m
.TH NGSRAT 3NCARG "October 1996" UNIX "NCAR GRAPHICS"
.na
.nh
.SH NAME
NGSRAT - save/restore/set NCAR GKS primitive attribute values in toto.
.SH SYNOPSIS
CALL NGSRAT(IOPT, IAT, RAT)
.SH C-BINDING SYNOPSIS
#include <ncarg/ncargC.h>
.sp
void c_ngsrat(int iopt, int *iat, float *rat)
.SH DESCRIPTION 
.IP IOPT 12
(an input variable of type INTEGER) indicating the desired action:
.IP "              = 0" 20
save the current settings of all attributes
internal to the subroutine, i.e. not in the IAT and RAT arrays.
.IP "              = 1" 20
restore the context from the most recently saved state.
.IP "              = 2" 20
return the current context in the output arrays IAT (integer 
attributes) and RAT (floating-point attributes) as follows:
.sp
IAT contains integer GKS attribute settings as per:
.IP " " 22
IAT( 1) = Clip indicator
.IP " " 22
IAT( 2) = Line type
.IP " " 22
IAT( 3) = Polyline color index
.IP " " 22
IAT( 4) = Marker type
.IP " " 22
IAT( 5) = Polymarker color index
.IP " " 22
IAT( 6) = Text font
.IP " " 22
IAT( 7) = Text precision
.IP " " 22
IAT( 8) = Text color index
.IP " " 22
IAT( 9) = Text path
.IP " " 22
IAT(10) = Text horizontal alignment
.IP " " 22
IAT(11) = Text vertical alignment
.IP " " 22
IAT(12) = Fill area interior style
.IP " " 22
IAT(13) = Fill are style index
.IP " " 22
IAT(14) = Fill area color index
.IP " " 20
RAT contains REAL GKS attribute settings as per:
.IP " " 22
RAT( 1) = Linewidth scale factor
.IP " " 22
RAT( 2) = Marker scale factor
.IP " " 22
RAT( 3) = Character expansion factor
.IP " " 22
RAT( 4) = Character spacing
.IP " " 22
RAT( 5) = Character height in world coordinates
.IP " " 22
RAT( 6) = Character up vector, X component in world coordinates
.IP " " 22
RAT( 7) = Character up vector, Y component in world coordinates
.IP "              = 3" 20
set the context to the values specified in the IAT and RAT arrays 
(as described above).
.IP IAT 12
An integer array, of length 14, that has meaning only if IOPT 
equals 2 or 3. 
The elements of the IAT array are described in the
above description of IOPT.
.IP RAT 12
A real array, of length 7, that has meaning only if IOPT 
equals 2 or 3.
The elements of the IAT array are described in the
above description of IOPT.
.SH USAGE
If you want to save the state of the NCAR GKS primitive attribute
settings, call NGSRAT with a first argument of 0.  When the first
argument is 0, the IAT and RAT arrays (second and third arguments)
are ignored and the state values are stored internally to NGSRAT.
.sp
To restore the NCAR GKS primitive attribute values to what they
were as of the last time NGSRAT was called with argument IOPT = 0,
then call NGSRAT with first argument of 1.  IAT and RAT are ignored
in this case.
.sp
If you call NGSRAT with a fist argument of 2, then the current
settings of the NCAR GKS primitive attribute values are returned 
in arrays IAT and RAT as per the description of IAT and RAT above.
.sp
If you call NGSRAT with a fist argument of 3, then arrays IAT and
RAT must be supplied and the values therein will be used to set 
the NCAR GKS primitive attribute values as per the description of
IAT and RAT above.
.sp
A common use of NGSRAT would be in conjunction with NGMFTC and
NGREOP for saving the GKS state after writing to one metafile, writing
to a new metafile, and then restoring the GKS state when reopening
the initial metafile.
.SH EXAMPLES
The following sequence:
.nf

        CALL NGSRAT(0,IAT,RAT)

          ... do some setting and drawing

        CALL NGSRAT(1,IAT,RAT)

.fi
would save the values of all NCAR GKS primitive attributes 
before the code that does setting and drawing and then restore the
attribute values after the setting and drawing.
.sp
Use the ncargex command to see the following relevant example: 
pgkex27.
.SH ACCESS
To use NGSRAT or c_ngsrat, load the NCAR Graphics libraries ncarg, ncarg_gks,
and ncarg_c, preferably in that order.  
.SH MESSAGES
NGSRAT issues no messages apart from potential illegal setting of
the attribute values when IOPT = 3.
.SH SEE ALSO
Online:
ngmftc(3NCARG),
ngreop(3NCARG),
.sp
Hardcopy:
NCAR Graphics Fundamentals, UNIX Version.
.SH COPYRIGHT
Copyright (C) 1987-2009
.br
University Corporation for Atmospheric Research
.br
The use of this Software is governed by a License Agreement.
